### ISSUE TITLE
sngrep localhost centos lost packets

### ISSUE THREAD

**Issue by luciocosmo** [irontec/sngrep#246]
Hi.

We're using sngrep from many months on CentOS (VirtualBox machine hosted on OSX or Windows) and OSX for development sip flow checking.

We are experiencing missing parts of flows display ONLY in CentOS localhost situations, where the packet transmitter and packet receiver are on the same CentOS machine. 

Packets are randomly lost, we can repeat this anytime also on newly installed VMs. 

When packets are transmitted from 2 distinct centos virtual machines or OSX to CentOS there are no packets lost and sip messages are fully visible and displayed.

Thanks in advance for any advice.
Regards.

**Comment by Kaian** [comment]
Hi @luciocosmo 

Are you able to reproduce this problem with a captured pcap from a raw capturing tool? Try running ngrep or tcpdump at the same time, and when you find something is missing check the pcap file with sngrep. If this is a parsing bug, having a test pcap helps a lot!

Are you using bpf filtering in command line? What transport are you capturing (TCP/UDP)?

Thanks for reporting!! 
Regards

**Comment by luciocosmo** [comment]
> Il giorno 22 mag 2018, alle ore 10:54, Kaian <notifications@github.com> ha scritto:
> 
> Hi @luciocosmo <https://github.com/luciocosmo>
> Are you able to reproduce this problem with a captured pcap from a raw capturing tool? Try running ngrep or tcpdump at the same time, and when you find something is missing check the pcap file with sngrep. 
> 
tcpdump -i any -w trace.pcap
sngrep -I trace.pcap



sngrep (no params, no configuration)




> If this is a parsing bug, having a test pcap helps a lot!
> 
> Are you using bpf filtering in command line? 
> 
no
> What transport are you capturing (TCP/UDP)?
> 
udp only
> Thanks for reporting!
> 
Thanks for so fast reply :)

L

**Comment by Kaian** [comment]
Can you attach or send me an email with a pcap file without sensitive data to debug why those messages doesn't match the regular expressions used to parse SIP packets? It helps a lot!.

Thanks!

**Comment by dovi5988** [comment]
We are having the exact same issue with physical hardware. For instance we have a SUBSCRIBE. With sngrep we see:

1) SUBSCRIBE from client to OpenSipS
2) 407 from OpenSipS to Client.
3) SUBSCRIBE with no-once from client.

with tcpdump we see:
1) SUBSCRIBE from client to OpenSipS
2) 407 from OpenSipS to Client.
3) SUBSCRIBE with no-once from client.
4) Response from OpenSipS 404 Not found

So OpenSipS is sending a 404 but sngrep isn't showing it, this randomly happens. There is sensitive data so I can not post them here publicly. If there is an email I can share the pcap's and screenshots privately.

**Comment by Kaian** [comment]
Send the test pcap to kaian@irontec.com and I'll check what's going on as soon as I can.

Thanks!

**Comment by luciocosmo** [comment]
Hi

I did not reply before, still waiting for company permission to publish the pcap. It contains reserved information.
We still see and experience the problem.

Also, is there a possibility to show "any sip-like" message, for instance writing ENERGY instead of REGISTER, something like a first header line ENERGY sip://abcd@efg.net SIP/2.0

It could be a lot useful.

Lucio

**Comment by Kaian** [comment]
You can send it to my mail. I'll remove them after debugging. 

For now, we try to keep strict SIP parsing in sngrep, so we stick to RFC 3261 for message format. That means each row must end with /r/n, methods are limited and headers must have the expected format.

Thanks!

**Comment by dovi5988** [comment]
@Kaian ,

I have just sent some sample PCAP's. What I did notice is that if I just did "sngrep" then we saw dropped packets. However if I specified ports (e.g. "sngrep port 5060 or port 5061") then there seemed to have been no dropped packets. I am not sure if this is the same issue that @luciocosmo is having or not.

**Comment by Kaian** [comment]
That's quite strange. Usually the problems may happen when port filters are used because some IP fragmented packed packets are not passed from libpcap to the application, but you have the oppossite.

I had a quick debug to the pcaps you sent me. First of all, thanks for the testing data and detailed information! Sadly, they didn't help me debugging the problem, because I was expecting sngrep failing while parsing tcpdump pcaps and that didn't happen.

I see there are some dialogs missing in sngrep captures, but they are parsed correctly from tcpdump captures, so this cann't be a SIP parser bug...

I'm planning some stress test to check if high traffic can affect the capture process.

Thanks!!

**Comment by dovi5988** [comment]
@Kaian not a problem. Let me know if there is anything else that I can help with. The sniffing was done on a box that has very little traffic (under 1 mbit).

**Comment by Kaian** [comment]
As mentioned in my previous note I've been doing some stress tests and have also suffered this problem, so now I have a place to debug what's going on. 

My guess is still related to capturing and parsing RTP packets (I think sngrep is not fast enough at relating each rtp packet with its call) and some kind of internal buffer in libpcap. If we are too slow and the buffer fills, it may skip some packets. I have no tested this, but could explain why using bpf filter (reducing the traffic to just SIP) will improve the performance and no packet is skipped.

Stay tuned!

**Comment by Kaian** [comment]
From [libpcap man pages](https://www.tcpdump.org/manpages/pcap.3pcap.html)

```
buffer size

Packets that arrive for a capture are stored in a buffer, so that they 
do not have to be read by the application as soon as they arrive. On 
some platforms, the buffer's size can be set; a size that's too small 
could mean that, if too many packets are being captured and the 
snapshot length doesn't limit the amount of data that's buffered, packets
could be dropped if the buffer fills up before the application can read 
packets from it, while a size that's too large could use more non-pageable 
operating system memory than is necessary to prevent packets from 
being dropped. The buffer size is set with pcap_set_buffer_size(). 
```

**Comment by luciocosmo** [comment]
Is this a sngrep configuration ?

**Comment by Kaian** [comment]
Nope, but it make sense with what I've said in the previous note. The problem could be related with slow parsing packets may lead to dropped packets.

I haven't checked your pcap, but could you try using bpf filters to reduce the packets parsed (like `sngrep port 5060`) and check if there are missing packets?

Thanks!

**Comment by luciocosmo** [comment]
> Il giorno 07 giu 2018, alle ore 12:48, Kaian <notifications@github.com> ha scritto:
> 
> Nope, but it make sense with what I've said in the previous note. The problem could be related with slow parsing packets may lead to dropped packets.
> 
> I haven't checked your pcap, but could you try using bpf filters to reduce the packets parsed (like sngrep port 5060) and check if there are missing packet
> 
> 

I'm lost with this statement.
Can you do an example ?

Thanks

**Comment by Kaian** [comment]
Could you reproduce this problem if you run sngrep using filters from command line?
For example, to just capturing packets that have as source or destination port 5060 or 5072 run from your shell:

```
sngrep port 5060 or port 5072
``` 

Thanks!

**Comment by luciocosmo** [comment]
> Il giorno 07 giu 2018, alle ore 21:11, Kaian <notifications@github.com> ha scritto:
> 
> Could you reproduce this problem if you run sngrep using filters from command line?
> For example, to just capturing packets that have as source or destination port 5060 or 5072 run from your shell:
> 
> sngrep port 5060 or port 5072

Ah now I got it. 

The low level capture gets ALL the packets from the network interface and sngrep has to filter sip from everything else ?
I also understand why the buffer must be wide enough. And makes sense: in these situations we have also an ssh on sngrep and/or more terminal open with lot of logs, so lots of network traffic (even more than the sip one is generated).

Correct ?

Lucio

**Comment by Kaian** [comment]
> The low level capture gets ALL the packets from the network interface and sngrep has to filter sip from everything else ?

Yes, that correct. Usually SIP is a small number of packets compared with RTP, so the fix for this requires improving parsing logics for SIP/RTP so we can keep up on high network traffic.

**Comment by luciocosmo** [comment]
Hi, after some time this is still open :) We're really using a lot this tool and enjoyed last upgrade which recognizes other sip methods and so on. 

By the way, our life would be much easier if there could be a parameter to trim the buffering and not to loose packets on centos, as above. Can I add an improvement request about ?

Or, can we recompile it with a bugger buffer fixed on centos ? Can you address me on the source position ?

Many thanks

**Comment by Kaian** [comment]
Hi!

We can increase libpcap buffer size using [pcap_set_buffer_size](http://www.tcpdump.org/manpages/pcap_set_buffer_size.3pcap.html) but I don't think it will fix the problem. After some time the buffer will be full and new packets will be skipped. You'll have the same problem and more memory used.

This real problem in my opinion is that sngrep actually parses, sorts and groups data. It does quite some process for each captured packet, contrary to other tools like tcpdump, ngrep, and so on. That makes sngrep not the best tool for high traffic captures. I've been working on a new sngrep version with all code refactored, using hash tables for matching all packets, and even with that, it will never be as fast a raw capture tool.

There are workarounds, like capturing using BPF filters to only match interesting packets (never run sngrep without bpf filters in a high traffic box), or capturing to a file with a raw tool and load it with sngrep. The buffer thing should not affect at all for offline captures from .pcap.

As said, we can add a buffer setting (like tcpdump -B flag) but I don't think that will solve the problem on the long run.

Regards!

**Comment by Kaian** [comment]
Hi!

I have created a new branch called `capture-buffer` to try to test this before creating a parameter and or configuration setting.

The above commit changes the capture creation to include pcap_set_buffer_size (it was not just a line of code in the end) with default value of 512K. This value can be changed updating [MAXIMUM_BUFFERSIZE](https://github.com/irontec/sngrep/blob/capture-buffer/src/capture.h#L90) define.

There is no actual function to retrieve set buffer size and I have no way to test if this actually improves performance, so It would be great if you could try that branch on your centos machine and play a bit with the define value. If you need help on compiling a specific branch, feel free to drop me a line and I'll try to provide instructions.

Regards!
